tests/unit/s2n_client_hello_test.c

```diff
@@ -1953,6 +1961,138 @@ int main(int argc, char **argv)
+        EXPECT_NOT_NULL(server_config);
+
+        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(server_config, chain_and_key));
+
+        struct s2n_cipher_suite *test_cipher_suites[REDUCED_CIPHER_SUITE_COUNT] = { 0 };
+
+        for (int i = 0; i < REDUCED_CIPHER_SUITE_COUNT; i++) {
+            test_cipher_suites[i] = &s2n_rsa_with_aes_128_gcm_sha256;
+        }
+
+        const struct s2n_cipher_preferences test_cipher_suites_preferences = {
+            .count = s2n_array_len(test_cipher_suites),
+            .suites = test_cipher_suites,
+        };
+
+        struct s2n_security_policy test_security_policy = {
+            .minimum_protocol_version = S2N_TLS12,
+            .cipher_preferences = &test_cipher_suites_preferences,
+            .kem_preferences = &kem_preferences_null,
+            .signature_preferences = &s2n_signature_preferences_20240501,
+            .ecc_preferences = &s2n_ecc_preferences_20240501,
+            .rules = {
+                    [S2N_PERFECT_FORWARD_SECRECY] = true,
+            },
+        };
+
+        EXPECT_SUCCESS(s2n_config_disable_x509_verification(client_config));
+
+        DEFER_CLEANUP(struct s2n_connection *client = s2n_connection_new(S2N_CLIENT),
+                s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(client);
+        EXPECT_SUCCESS(s2n_connection_set_config(client, client_config));
+        EXPECT_SUCCESS(s2n_connection_set_blinding(client, S2N_SELF_SERVICE_BLINDING));
+        client->security_policy_override = &test_security_policy;
+
+        DEFER_CLEANUP(struct s2n_connection *server = s2n_connection_new(S2N_SERVER),
+                s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(server);
+        EXPECT_SUCCESS(s2n_connection_set_config(server, server_config));
+        EXPECT_SUCCESS(s2n_connection_set_blinding(server, S2N_SELF_SERVICE_BLINDING));
+        server->security_policy_override = &test_security_policy;
+
+        DEFER_CLEANUP(struct s2n_test_io_stuffer_pair io_pair = { 0 }, s2n_io_stuffer_pair_free);
+        EXPECT_OK(s2n_io_stuffer_pair_init(&io_pair));
+        EXPECT_OK(s2n_connections_set_io_stuffer_pair(client, server, &io_pair));
+
+        EXPECT_OK(s2n_negotiate_test_server_and_client_until_message(server, client, SERVER_HELLO));
+
+        /* handshake.io shouldn't be tainted after sending and receiving large client hello */
+        EXPECT_TRUE(client->handshake.io.tainted == 0);
+        EXPECT_TRUE(server->handshake.io.tainted == 0);
+
+        struct s2n_client_hello *client_hello = s2n_connection_get_client_hello(server);
+        EXPECT_NOT_NULL(client_hello);
+        uint32_t handshake_max_len_margin = S2N_MAXIMUM_HANDSHAKE_MESSAGE_LENGTH - s2n_client_hello_get_raw_message_length(client_hello);
+        /* Size of Client Hello is less than 300 bytes from the maximum handshake length */
+        EXPECT_TRUE(handshake_max_len_margin < ESTIMATED_MAX_HANDSHAKE_LENGTH_MARGIN);
```

Comment:
This seems unnecessarily complex to me. Why not just verify that the size of the client hello is <64k?